### ISSUE TITLE
 X atom types should be an enum that can be converted to a &str

### ISSUE THREAD

**Issue by sminez** [sminez/penrose#41]
Currently, the X atoms are held as a simple slice of static strings that are then referenced an copied in multiple places. These should really be an enum that can be converted to a `&str` in a similar way to `WindowType` in the `draw` module:

```rust
    /// An EWMH Window type
    pub enum WindowType {
        /// A dock / status bar
        Dock,
        /// A menu
        Menu,
        /// A normal window
        Normal,
    }
    impl WindowType {
        pub(crate) fn as_ewmh_str(&self) -> &str {
            match self {
                WindowType::Dock => "_NET_WM_WINDOW_TYPE_DOCK",
                WindowType::Menu => "_NET_WM_WINDOW_TYPE_MENU",
                WindowType::Normal => "_NET_WM_WINDOW_TYPE_NORMAL",
            }
        }
    }
```

This would also require altering struct / function definitions to use the new enum and updating existing uses in the code base.

**Comment by Danacus** [comment]
Hi, it's me again. This looked like it would be some fun vim practice. Here's what I have so far:

 ```rust
/**
 * Enum with X Atoms. See 'WindowType' for '_NET_WINDOW_TYPE' atoms.
 */
pub enum Atom {
    /// MANAGER
    Manager,
    /// UTF8_STRING
    UTF8String,
    /// WM_CLASS
    WmClass,
    /// WM_DELETE_WINDOW
    WmDeleteWindow,
    /// WM_PROTOCOLS
    WmProtocols,
    /// WM_STATE
    WmState,
    /// WM_NAME
    WmName,
    /// WM_TAKE_FOCUS
    WmTakeFocus,
    /// _NET_ACTIVE_WINDOW
    NetActiveWindow,
    /// _NET_CLIENT_LIST
    NetClientList,
    /// _NET_CURRENT_DESKTOP
    NetCurrentDesktop,
    /// _NET_DESKTOP_NAMES
    NetDesktopNames,
    /// _NET_NUMBER_OF_DESKTOPS
    NetNumberOfDesktops,
    /// _NET_SUPPORTED
    NetSupported,
    /// _NET_SUPPORTING_WM_CHECK
    NetSupportingWmCheck,
    /// _NET_SYSTEM_TRAY_OPCODE
    NetSystemTrayOpcode,
    /// _NET_SYSTEM_TRAY_ORIENTATION
    NetSystemTrayOrientation,
    /// _NET_SYSTEM_TRAY_ORIENTATION_HORZ
    NetSystemTrayOrientationHorz,
    /// _NET_SYSTEM_TRAY_S0
    NetSystemTrayS0,
    /// _NET_WM_DESKTOP
    NetWmDesktop,
    /// _NET_WM_NAME
    NetWmName,
    /// _NET_WM_STATE
    NetWmState,
    /// _NET_WM_STATE_FULLSCREEN
    NetWmStateFullscreen,
    /// _NET_WM_WINDOW_TYPE
    NetWmWindowType,
    /// _XEMBED
    XEmbed,
    /// _XEMBED_INFO
    XEmbedInfo,
}

/**
 * Enum with '_NET_WM_WINDOW_TYPE' atoms.
 */
pub enum WindowType {
    /// _NET_WM_WINDOW_TYPE_DESKTOP
    Desktop,
    /// _NET_WM_WINDOW_TYPE_DOCK
    Dock,
    /// _NET_WM_WINDOW_TYPE_TOOLBAR
    Toolbar,
    /// _NET_WM_WINDOW_TYPE_MENU
    Menu,
    /// _NET_WM_WINDOW_TYPE_UTILITY
    Utility,
    /// _NET_WM_WINDOW_TYPE_SPLASH
    Splash,
    /// _NET_WM_WINDOW_TYPE_DIALOG
    Dialog,
    /// _NET_WM_WINDOW_TYPE_DROPDOWN_MENU
    DropdownMenu,
    /// _NET_WM_WINDOW_TYPE_POPUP_MENU
    PopupMenu,
    /// _NET_WM_WINDOW_TYPE_NOTIFICATION
    Notification,
    /// _NET_WM_WINDOW_TYPE_COMBO
    Combo,
    /// _NET_WM_WINDOW_TYPE_DND
    Dnd,
    /// _NET_WM_WINDOW_TYPE_NORMAL
    Normal,
}

const AUTO_FLOAT_WINDOW_TYPES: &[WindowType] = &[
    WindowType::Desktop,
    WindowType::Dialog,
    WindowType::Dock,
    WindowType::DropdownMenu,
    WindowType::Menu,
    WindowType::Notification,
    WindowType::PopupMenu,
    WindowType::Splash,
    WindowType::Toolbar,
    WindowType::Utility,
];
```

I decided to split Window Type atoms in their own enum. I wanted to know what you think before I start replacing all uses.
Also: should I `use Atom::*` or should I write `Atom::xxx` on every occurrence?

**Comment by Danacus** [comment]
On second though it might be easier to write a declarative macro, because we would also need a conversion method and a slice with all variants. So adding or removing an atom would require changing it in 3 places.

**Comment by sminez** [comment]
Hey @Danacus ! 👋 
What you've got so far is looking good yeah. I agree that a macro is probably going to be the way to go with this. I haven't used it before but [strum](https://crates.io/crates/strum_macros) might be worth a look? (I've been meaning to work through the proc-macro workshop stuff to deal with this sort of thing directly but I haven't had any time to get started with it yet. That said, strum seems to provide exactly what we'd want.)

In terms of usage, I'd prefer your second suggestion (i.e. fully qualified) as it makes things easier for people to grep for in the code base 🙂

**Comment by Danacus** [comment]
I saw strum being mentioned quite often on other places, but I was able to write the following declarative macro that does the trick:

```rust
#[macro_export]
macro_rules! gen_atoms {
    ($enum_name:ident, $slice_name:ident, { $([$variant:ident, $atom:expr]),+ $(,)? }) => {
        #[derive(Clone, Hash, PartialEq, Eq)]
        pub enum $enum_name {
            $(
            #[doc = $atom]
            $variant,
            )+
        }

        impl $enum_name {
            pub(crate) fn as_atom_str(&self) -> &str {
                match self {
                    $($enum_name::$variant => $atom,)+
                }
            }
        }

        const $slice_name: &[$enum_name] = &[
            $($enum_name::$variant,)+
        ];
    };
}
```

This way we can also generate a `Atom` to `&str` method and have documention that shows the original atom name. 

It can be used like this:
```rust
gen_atoms!(Atom, ATOMS, {
    [Manager, "MANAGER"],
    [UTF8String, "UTF8_STRING"],
    [WmClass, "WM_CLASS"],
    [WmDeleteWindow, "WM_DELETE_WINDOW"],
    [WmProtocols, "WM_PROTOCOLS"],
    [WmState, "WM_STATE"],
    [WmName, "WM_NAME"],
    [WmTakeFocus, "WM_TAKE_FOCUS"],
    [NetActiveWindow, "_NET_ACTIVE_WINDOW"],
    [NetClientList, "_NET_CLIENT_LIST"],
    [NetCurrentDesktop, "_NET_CURRENT_DESKTOP"],
    [NetDesktopNames, "_NET_DESKTOP_NAMES"],
    [NetNumberOfDesktops, "_NET_NUMBER_OF_DESKTOPS"],
    [NetSupported, "_NET_SUPPORTED"],
    [NetSupportingWmCheck, "_NET_SUPPORTING_WM_CHECK"],
    [NetSystemTrayOpcode, "_NET_SYSTEM_TRAY_OPCODE"],
    [NetSystemTrayOrientation, "_NET_SYSTEM_TRAY_ORIENTATION"],
    [NetSystemTrayOrientationHorz, "_NET_SYSTEM_TRAY_ORIENTATION_HORZ"],
    [NetSystemTrayS0, "_NET_SYSTEM_TRAY_S0"],
    [NetWmDesktop, "_NET_WM_DESKTOP"],
    [NetWmName, "_NET_WM_NAME"],
    [NetWmState, "_NET_WM_STATE"],
    [NetWmStateFullscreen, "_NET_WM_STATE_FULLSCREEN"],
    [NetWmWindowType, "_NET_WM_WINDOW_TYPE"],
    [XEmbed, "_XEMBED"],
    [XEmbedInfo, "_XEMBED_INFO"],
});

gen_atoms!(WindowType, WINDOW_TYPES, {
    [Desktop, "_NET_WM_WINDOW_TYPE_DESKTOP"],
    [Dock, "_NET_WM_WINDOW_TYPE_DOCK"],
    [Toolbar, "_NET_WM_WINDOW_TYPE_TOOLBAR"],
    [Menu, "_NET_WM_WINDOW_TYPE_MENU"],
    [Utility, "_NET_WM_WINDOW_TYPE_UTILITY"],
    [Splash, "_NET_WM_WINDOW_TYPE_SPLASH"],
    [Dialog, "_NET_WM_WINDOW_TYPE_DIALOG"],
    [DropdownMenu, "_NET_WM_WINDOW_TYPE_DROPDOWN_MENU"],
    [PopupMenu, "_NET_WM_WINDOW_TYPE_POPUP_MENU"],
    [Notification, "_NET_WM_WINDOW_TYPE_NOTIFICATION"],
    [Combo, "_NET_WM_WINDOW_TYPE_COMBO"],
    [Dnd, "_NET_WM_WINDOW_TYPE_DND"],
    [Normal, "_NET_WM_WINDOW_TYPE_NORMAL"],
});

```

What do you think about this?

**Comment by sminez** [comment]
ah cool, looks good :+1: 
Other than renaming the macro to something more general (it's broader than just making the `atoms` slice it seems?) that should be good! You should be able to extend the macro to allow for adding doc comments to the generated enum and slice which would be good because these are going to show up in a variety of places I suspect. (see [this](https://github.com/sminez/arthroprod/blob/master/src/prelude.rs#L8) macro I have in another project for an example)

**Comment by sminez** [comment]
implemented in #43
